api/v1alpha1/stage_types.go

```diff
@@ -200,16 +207,50 @@ type StageSubscription struct {
 // FreightRequest expresses a Stage's need for Freight having originated from a
 // particular Warehouse.
 type FreightRequest struct {
-	// Origin specifies what Warehouse the requested Freight must have originated
-	// from. This is a required field.
+	// Origin specifies from where the requested Freight must have originated.
+	// This is a required field.
 	//
 	// +kubebuilder:validation:Required
-	Origin string `json:"origin" protobuf:"bytes,1,opt,name=origin"`
+	Origin FreightOrigin `json:"origin" protobuf:"bytes,1,opt,name=origin"`
 	// Sources describes where the requested Freight may be obtained from. This is
 	// a required field.
 	Sources FreightSources `json:"sources" protobuf:"bytes,2,opt,name=sources"`
 }
 
+// FreightOrigin describes a kind of Freight in terms of where it may have
+// originated.
+//
+// +protobuf.options.(gogoproto.goproto_stringer)=false
+type FreightOrigin struct {
+	// Kind is the kind of resource from which Freight may have originated. At
+	// present, this can only be "Warehouse".
+	//
+	// +kubebuilder:validation:Required
+	Kind FreightOriginKind `json:"kind" protobuf:"bytes,1,opt,name=kind"`
+	// Name is the name of the resource of the kind indicated by the Kind field
+	// from which Freight may originated.
+	//
+	// +kubebuilder:validation:Required
+	Name string `json:"name" protobuf:"bytes,2,opt,name=name"`
+}
+
+func (f *FreightOrigin) String() string {
+	if f == nil {
+		return ""
+	}
+	return fmt.Sprintf("%s[%s]", f.Kind, f.Name)
```

Comment:
I kind of feel something like

```suggestion
	return fmt.Sprintf("%s/%s", f.Kind, f.Name)
```

might be a better fit. As `[]` is often used for grouping (e.g. API group versions, see `GroupVersions` in Kubernetes' `schema` package), while `/` is a common way for GVK / name joins in the Kubernetes (ecosystem) itself.

Reply:
What I had been going for was making it feel like it identifies a specific origin from a collection of origins, but I am equally happy with the alternative you have suggested. Happy to change.